### ISSUE TITLE
ImGui::BeginViewportSideBar() and moving windows

### ISSUE THREAD

**Issue by KennyProgrammer** [ocornut/imgui#7118]
**Version/Branch of Dear ImGui:**

Version: 1.88
Branch: docking

**Back-end/Renderer/Compiler/OS**

Back-ends imgui_impl_win32.cpp + imgui_impl_opengl3.cpp/ imgui_impl_dx11.cpp
Compiler: MSVC
Operating System: Windows 10

**My Issue/Question:**
When I try to move a window through the sidebar (aka `ImGui::BeginViewportSideBar()` ), just as I move a window to any area of the main window, the position of the window being moved to the border of the sidebar. This happens only when I create it with UP direction, when I create a sidebar, for example, for a status bar with the DOWN direction, there is nothing like that, everything works fine.

Also when **popups, tooltips or any window** are opened on sidebar, it shifts its position to dockspace window (e.g. below all sibebars).

**Standalone, minimal, complete and verifiable example:** 

Here example how i create tool bar using sidebar:

```
ImGuiViewportP* viewport = (ImGuiViewportP*)(void*)ImGui::GetMainViewport();
ImGuiWindowFlags windowFlags = ImGuiWindowFlags_NoScrollbar | ImGuiWindowFlags_NoScrollWithMouse | ImGuiWindowFlags_NoSavedSettings | ImGuiWindowFlags_NoNavFocus;

// Toolbar Window (Panel) uses as BeginViewportSideBar.
ImGui::BeginViewportSideBar("##ToolBar", viewport, ImGuiDir_Up, 28.0f, windowFlags | ImGuiWindowFlags_NoNavInputs);

// ... render some content here...

ImGui::End();
```

**Screenshots/Video**

Here a few videos how what problem looks like:

In first video firstly i show how it should be and then when i move window into main window the window doesnt move above sidebar.

https://github.com/ocornut/imgui/assets/56895345/263d4cf1-4718-4e3a-aecb-653a31f43b0e

And the same with another windows (tooltips, popups, etc)

https://github.com/ocornut/imgui/assets/56895345/cefe1bb6-a276-489f-a933-10ee6075df34

PS: I know that `ImGui::BeginViewportSideBar()` API is beta and not public, so its actually bug of this function or problem because it called `BeginViewport `and work with viewports not with windows? So if it is, the are exist alternative ways to create toolbars? 

PSS: This is not critical but really annoying that you can move windows through sidebars.

**Comment by ocornut** [comment]
Do you have `io.ConfigWindowsMoveFromTitleBarOnly` set to true?

If you can only move a window by its title bar we make it so that title bar can never be over a side bar (e.g. main menu bar). Otherwise we would need to guarantee that your window can never be covered by the side bar.

Perhaps we can have a system where if you specify `ImGuiWindowFlags_NoBringToFrontOnFocus` for the side bar then its geometry is declared differently in viewport->WorkRect.

If you don't call `BeginViewportSideBar()` more than once, you can backup and restore `viewport->BuildWorkOffsetMin` and `viewport->BuildWorkOffsetMax` in order to not alter the viewport WorkRect. Or you may do this directly:

```cpp
ImGui::SetNextWindowPos(viewport->Pos);
ImGui::SetNextWindowSize(viewport->Size);
ImGui::Begin(....);
```

**Comment by KennyProgrammer** [comment]
> Do you have io.ConfigWindowsMoveFromTitleBarOnly set to true?

Yes, I have this flag set, I put it a long time ago, it seems to have solved one of my problems earlier. I'll try to remove it and see if something else breaks.

**After test:** This actually fix problem on 50%, now another issue happens that when this window intersects with platform non-client area that i defined for my main window, i cannot move that window, because it acts as non-client and move main window instead.  Perhaps this can be resolved in the same way as I blocked non-client area interaction for any widget inside the title bar that is located in the non-client area. (e.g. by returning client area instead in windows code when `ImGui::IsAnyItemActive()`)

> If you can only move a window by its title bar we make it so that title bar can never be over a side bar (e.g. main menu bar). **Otherwise we would need to guarantee that your window can never be covered by the side bar.**

I did not understand the last sentence a little, why not just let the window move freely (at least at the moment when it is inside the main window (because it does not use the platform window interface))

> Perhaps we can have a system where if you specify ImGuiWindowFlags_NoBringToFrontOnFocus for the side bar then its geometry is declared differently in viewport->WorkRect.

I noticed that I don't have this flag on the sidebar, I'll try to set it and maybe it will be a solution.

> If you don't call `BeginViewportSideBar() `more than once, you can backup and restore `viewport->BuildWorkOffsetMin` and `viewport->BuildWorkOffsetMax` in order to not alter the viewport `WorkRect`.

I call this more that once, for TitleBar, MenuBar, Toolbar and StatusBar.

**Comment by KennyProgrammer** [comment]
**So i finally resolve problem.**
For most cases and users that use ImGui setting `io.ConfigWindowsMoveFromTitleBarOnly` to `false` will be enough witch is by default.

But for my game engine case not. So i read code and found where `ConfigWindowsMoveFromTitleBarOnly` actually used.
 - First it used to **Cancel moving if clicked outside of title bar** this is exactly what this flag was originally intended for (I think 
   so) in the `UpdateMouseMovingWindowEndFrame` function.
 - Second it used in `ClampWindowRect` unfortunately, I do not fully understand why (it may solve some other problem). But this was actual problem in my case.
 
I don't like to modify the source code of any library, but I had to. At least until there is an alternative solution. So I created a new `ConfigWindowsMoveFromTitleBarOnlyEx` flag in **imgui.h** which I wrapped in the `FE_IMGUI_C0005` macro (yes, I have already modified ImGui 4 times, not significant changes, but significant for my engine) so that it could be turned off and it was easier for me to switch to the new version. And implemented only for the first case, ignoring the second. And set this flag to true insted original `ConfigWindowsMoveFromTitleBarOnly`, and it worked.

**Comment by ocornut** [comment]
> Second it used in ClampWindowRect unfortunately, I do not fully understand why (it may solve some other problem). But this was actual problem in my case.

Because otherwise your window can get stuck behind another window and you can never move it again is the only moving pointi s the title bar.

**Comment by KennyProgrammer** [comment]
> Because otherwise your window can get stuck behind another window and you can never move it again is the only moving pointi s the title bar.

Well, for this, as I understand it, there is already a check here so technically we shouldn't be in this situation.

`UpdateMouseMovingWindowEndFrame`

or rather, in these lines:
```
 if (g.IO.ConfigWindowsMoveFromTitleBarOnly)
     if (!(root_window->Flags & ImGuiWindowFlags_NoTitleBar) || root_window->DockIsActive)
         if (!root_window->TitleBarRect().Contains(g.IO.MouseClickedPos[0]))
             g.MovingWindow = NULL;
```

By the second case, I meant this function:

```
static inline void ClampWindowRect(ImGuiWindow* window, const ImRect& visibility_rect)
{
    ImGuiContext& g = *GImGui;
    ImVec2 size_for_clamping = window->Size;
    if (g.IO.ConfigWindowsMoveFromTitleBarOnly && (!(window->Flags & ImGuiWindowFlags_NoTitleBar) || window->DockNodeAsHost))
        size_for_clamping.y = ImGui::GetFrameHeight(); // Not using window->TitleBarHeight() as DockNodeAsHost will report 0.0f here.
    window->Pos = ImClamp(window->Pos, visibility_rect.Min - size_for_clamping, visibility_rect.Max);
}
```

Here when used `g.IO.ConfigWindowsMoveFromTitleBarOnly`, it set `size_for_clamping.y` as `ImGui::GetFrameHeight()` of the window i guess.

Exactly when I originally had ConfigWindowsMoveFromTitleBarOnly to true it calls this:
`size_for_clamping.y = ImGui::GetFrameHeight();`

and then i cannot move window througth sidebar.

**Comment by ocornut** [comment]
You misunderstood my message, read it again. Without this position clamping code (the second part) it would become possible to "lose" a window behind a sidebar and never be able to move it again.

The only valid case where it would be acceptable to alter the clamping logic is if we can guarantee that the sidebar will never be brought to front on focus.

**Comment by KennyProgrammer** [comment]
Now I got it. It's just that I've been testing it and I've never had a case where the sidebar came to the fore. And then I saw that I already had the `ImGuiWindowFlags_NoBringToFrontOnFocus` flag set on the sidebar which guarantee that the sidebar will never be brought to front on focus.

Sorry, for misunderstanding.

**Comment by KennyProgrammer** [comment]
Like this: 
![image](https://github.com/ocornut/imgui/assets/56895345/7a8b16ac-cdbe-4d22-a19e-ec486c145423)
And i guess if i make status bar even bigger and without `ImGuiWindowFlags_NoBringToFrontOnFocus` that will be what you talking about.

**Comment by KennyProgrammer** [comment]
Exactly, and especially when `ConfigWindowsMoveFromTitleBarOnly` is set to true, so as not to create such minor problems.

Anyways, thank you :)

**Comment by MirrasHue** [comment]
Hey @ocornut / @GamingMinds-DanielC, is there a way to use `ConfigWindowsMoveFromTitleBarOnly` for some specific windows but not for all? I tried enabling it before said windows' and disabling it after in the same frame, but it didn't work. And unfortunately there's no `ImGuiWindowFlags_` that address this for individual windows.

**Comment by ocornut** [comment]
@MirrasHue Hello, could you describe your concrete use case that would require this? It seems like it would make interactions inconsistent for the end user.

**Comment by MirrasHue** [comment]
@ocornut At the moment I'm building a simple 2D game engine, and it supports multiple scenes (ImGui viewport) opened at once. I didn't want to impose on the user to have them docked all the time, so we can undock them (to see more than one scene at the same time, e.g.). We generally don't want the viewport to move if the user is interacting with the scene, like clicking to select multiple entities, or dragging a gizmo (I found `ImGui::StopMouseMovingWindow()` for this case, maybe not ideal?). Therefore, it would be better if they are dragged only by the titlebar.

Now concerning the other panels (scene hierarchy, properties, content browser), it seems useful and handy to drag them by any empty space inside the window when they're not docked, that's why I was looking for a way to do this.

**Comment by ocornut** [comment]
This seems like a XY Problem then. You can catch clicks in the viewport by using eg an ImageButton() or InvisibleButton(). Please open a dedicated issue if you have more question about this.

**Comment by MirrasHue** [comment]
Indeed, just enabling `ConfigWindowsMoveFromTitleBarOnly` is more convenient. I guess I was so used to dragging windows's from empty space, that didn't want the feature to go away hahah

**Comment by ocornut** [comment]
This is not what I am saying and your answer suggest your misunderstood my post.

**Comment by MirrasHue** [comment]
Yeah, I know XD. It was a self reflection because I'll end up using that config.
